File: java/believe/datamodel/MutableValue.java
package believe.datamodel;

/**
 * A {@link MutableData} encapsulating a single value.
 *
 * @param <T> the type of the mutable value.
 */
public final class MutableValue<T> implements MutableData<T> {
  private T value;

  private MutableValue(T initialValue) {
    this.value = initialValue;
  }

  public static <T> MutableValue<T> of(T initialValue) {
    return new MutableValue<>(initialValue);
  }

  @Override
  public T get() {
    return value;
  }

  @Override
  public void update(T data) {
    value = data;
  }
}

File: java/believe/datamodel/MutableValue.kt
package believe.datamodel

/**
 * A [MutableData] encapsulating a single value.
 *
 * @param [T] the type of the mutable value.
 */
class MutableValue<T> private constructor(private var value: T) : MutableData<T> {
    override fun get(): T {
        return value
    }

    override fun update(data: T) {
        value = data
    }

    companion object {
        /**
         * Creates an instance of [MutableValue] using [initialValue] as its initial value.
         */
        @JvmStatic
        fun <T> of(initialValue: T): MutableValue<T> {
            return MutableValue(initialValue)
        }
    }
}
